### PR TITLE
[DDW-584] Apply withdrawals to Hardware Wallets

### PR DESCRIPTION
**Pull request by tomislavhoracek**
This PR introduces extended transaction signing / sending data with withdrawals with Hardware Wallets. Now we can use rewards along with Hardware wallets!

---

**NOTE: DO NOT MERGE**
- WBE bumped to the latest master where coin selection with withdrawals is included. We need to check this before merging @nikolaglumac  (REV: `b6810e8401b04b1d8a1a142f0b06b47cd6b51186`)

## Testing Checklist

- [Slack QA thread](https://input-output-rnd.slack.com/archives/GGKFXSKC6/p1614274478028600)
- [ ] Try to send a transaction with an amount that contains `available wallet balance` + `rewards`. Test with Ledger and Trezor
- [ ] Once you sent Transaction using withdrawals (rewards) try to create also a regular transaction without rewards just to ensure that TX Send is working for all cases

**Withdrawals Example:**

- You have a wallet "Wallet Test"
- Wallet balance is 50 ADA (shown in sidebar)
- You earned 2 ADA as rewards
- This means that you have `walletBalance = 50`, `availableWalletBalance = 48`, `rewards = 2`
- Try to send `49 ADA` - this means that you want to send your `full wallet available balance without rewards` + `rewards` to cover the transaction amount.
- Once you start a process, first indicator is to see `withdrawals` on your HW device. After that, once tx sent, you can see rewards sent in transaction details e.g. summary screen.
- e.g. if you have 3 ADA from rewards and e.g. you needed just 1 ADA to cover all tx requirements, the rest rewards are returned as a changed output to the wallet and are included in the wallet available balance which ( you will not have withdrawals anymore).

## Test Cases

```Gherkin
Scenario 1: Transaction on no reward wallet
Given I have a wallet with enough ADA for transaction fee
And the wallet doesn't have any reward
When I send some ADA
Then it will be successful

Scenario 2: Delegate on no reward wallet
Given I have a wallet with enough ADA for delegation
And the wallet doesn't have any reward
When I delegate the wallet
Then it will be successful

Note: I put numbers to make it easier to visualise
Scenario 3: Available balance, reward presence (Send less than available balance)
Given I have a wallet with enough ADA for transaction fee
And the wallet has available balance 10 ADA and reward 5 ADA
When I send 8 ADA
Then the transaction will be successful
And the available balance will be 7 ADA-transaction fees

Note: I put numbers to make it easier to visualise
Scenario 4: Available balance, reward presence (Send greater than available balance)
Given I have a wallet with enough ADA for transaction fee
And the wallet has available balance 10 ADA and reward 5 ADA
When I send 12 ADA
Then the transaction will be successful
And the available balance will be 3 ADA-transaction fees

Note: I put numbers to make it easier to visualise
Scenario 5: Delegate to a pool: Available balance, reward presence (Available balance less than 10 ADA)
Given I have a wallet with enough ADA for transaction fee
And the wallet has available balance 8 ADA and reward 5 ADA
When I attempt to delegate to a pool
Then I won't be able to

Note: I put numbers to make it easier to visualise
Scenario 6: Delegate to a pool: Available balance, reward presence (Available balance >=  10 ADA)
Given I have a wallet with enough ADA for transaction fee
And the wallet has available balance 10 ADA and reward 5 ADA
When I delegate to a pool
Then it will be successful
And the reward remains at 5 ADA

Scenario 7: Available balance 1 ADA, reward presence 3 ADA
Given I have a wallet with available balance 1 ADA and reward 3 ADA
When I send 1 ADA
Then the transaction will be successful
And the available balance will be 3 ADA-transaction fees


Still under construction
```
---

### Basics

- [ ] PR has been assigned and has appropriate labels (`feature`/`bug`/`chore`, `release-x.x.x`)
- [ ] PR is updated to the most recent version of the target branch (and there are no conflicts)
- [ ] PR has a good description that summarizes all changes
- [ ] PR has default-sized Daedalus window screenshots or animated GIFs of important UI changes:
  - [ ] In English
  - [ ] In Japanese
- [ ] CHANGELOG entry has been added to the top of the appropriate section (*Features*, *Fixes*, *Chores*) and is linked to the correct PR on GitHub
- [ ] Automated tests: All acceptance and unit tests are passing (`yarn test`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *development* build (`yarn dev`)
- [ ] Manual tests (minimum tests should cover newly added feature/fix): App works correctly in *production* build (`yarn package` / CI builds)
- [ ] There are no *flow* errors or warnings (`yarn flow:test`)
- [ ] There are no *lint* errors or warnings (`yarn lint`)
- [ ] There are no *prettier* errors or warnings (`yarn prettier:check`)
- [ ] There are no missing translations (running `yarn manage:translations` produces no changes)
- [ ] Text changes are proofread and approved (Jane Wild / Amy Reeve)
- [ ] Japanese text changes are proofread and approved (Junko Oda)
- [ ] UI changes look good in all themes (Alexander Rukin)
- [ ] Storybook works and no stories are broken (`yarn storybook`)
- [ ] In case of dependency changes `yarn.lock` file is updated

### Code Quality
- [ ] Important parts of the code are properly commented and documented
- [ ] Code is properly typed with flow
- [ ] React components are split-up enough to avoid unnecessary re-renderings
- [ ] Any code that only works in main process is neatly separated from components

### Testing
- [ ] New feature/change is covered by acceptance tests
- [ ] New feature/change is manually tested and approved by QA team
- [ ] All existing acceptance tests are still up-to-date
- [ ] New feature/change is covered by Daedalus Testing scenario
- [ ] All existing Daedalus Testing scenarios are still up-to-date

### After Review
- [ ] Merge the PR
- [ ] Delete the source branch
- [ ] Move the ticket to `done` column on the YouTrack board
- [ ] Update Slack QA thread by marking it with a green checkmark

### CODE REVIEW

File: nix/sources.json [hunk 1]
```diff
@@ -29,10 +29,10 @@
         "homepage": null,
         "owner": "input-output-hk",
         "repo": "cardano-wallet",
-        "rev": "a37c9856b4d96286e3f01a95026c63b986ebbba0",
-        "sha256": "1mg8n58j2mjqhhzjb4p5yp8z06b9arh40pagi9rddil2f3vxzihm",
+        "rev": "b6810e8401b04b1d8a1a142f0b06b47cd6b51186",
```
Comment (nikolaglumac) [inline nix/sources.json:32]: @tomislavhoracek please see if this version is enough: `b64b63fd56dddbe3ff070dea78977f0f15a487c2 `

That's the one we used here: https://github.com/input-output-hk/daedalus/pull/2292/files#diff-b9a0014cf6c4615d1c2e8d921dc3db59b4a68b462788d667ed622331660d17cd

File: CHANGELOG.md [hunk 2]
```diff
@@ -5,6 +5,7 @@ Changelog
 
 ### Chores
 
+- Withdrawals applied to the Hardware Wallets transactions ([PR 2408](https://github.com/input-output-hk/daedalus/pull/2408))
```
Comment (nikolaglumac) [inline CHANGELOG.md:8]: @tomislavhoracek please add this chagne under "Features" and use the following:

> Enabled rewards withdrawals on hardware wallets

**Review by yakovkaravelov — CHANGES_REQUESTED** [review]
@tomislavhoracek Great work. 🥇 
Left minor comments. Would you please take a look at them?

File: source/renderer/app/api/api.js [hunk 3]
```diff
@@ -1000,6 +1001,12 @@ export default class AdaApi {
         });
       }
 
+      const withdrawalsData = map(response.withdrawals, (withdrawal) => ({
```
[the remark this answers is not in the document]
Comment (tomislavhoracek) [reply source/renderer/app/api/api.js:1004]: @yakovkaravelov we are not using type annotations in `map` and `withdrawals` are already typed in top-level as a response from coin selection so  this is not necessary

File: source/renderer/app/api/api.js [hunk 4]
```diff
@@ -1000,6 +1001,12 @@ export default class AdaApi {
...
+        stakeAddress: withdrawal.stake_address,
+        derivationPath: withdrawal.derivation_path,
+        amount: withdrawal.amount,
+      }));
+
       const depositsArray = map(
         response.deposits,
         (deposit) => deposit.quantity
```
[the remark this answers is not in the document]
Comment (tomislavhoracek) [reply source/renderer/app/api/api.js:1012]: @yakovkaravelov same as above. Deposits are already typed through response with `CoinSelectionsResponse`

File: source/renderer/app/api/api.js [hunk 5]
```diff
@@ -1012,6 +1019,19 @@ export default class AdaApi {
         delegation && delegation.delegationAction === DELEGATION_ACTIONS.QUIT
           ? new BigNumber(DELEGATION_DEPOSIT).multipliedBy(LOVELACES_PER_ADA)
           : new BigNumber(0);
+
+      const withdrawalsArray = map(
+        response.withdrawals,
+        (withdrawal) => withdrawal.amount.quantity
```
[the remark this answers is not in the document]
Comment (tomislavhoracek) [reply source/renderer/app/api/api.js:1025]: @yakovkaravelov same as above. Already typed

File: source/renderer/app/stores/HardwareWalletsStore.js [hunk 6]
```diff
@@ -1346,19 +1366,24 @@ export default class HardwareWalletsStore extends Store {
...
+
       // Prepare unsigned transaction structure for serialzation
       const unsignedTx = prepareTxAux({
         txInputs: unsignedTxInputs,
         txOutputs: unsignedTxOutputs,
         fee,
         ttl,
         certificates: unsignedTxCerts,
-        withdrawals,
+        withdrawals: unsignedTxWithdrawals,
```
[the remark this answers is not in the document]
Comment (tomislavhoracek) [reply source/renderer/app/stores/HardwareWalletsStore.js:1384]: @yakovkaravelov done

**Review by yakovkaravelov — APPROVED** [review]

File: nix/sources.json (hunk 1 above)
Comment (tomislavhoracek) [reply nix/sources.json:32]: @nikolaglumac Withdrawals are not applied to coins selection from that revision.
Your CWB rev - `Feb 15, 2021`
Withdrawals applied in rev- `Feb 18, 2021`

File: CHANGELOG.md (hunk 2 above)
Comment (tomislavhoracek) [reply CHANGELOG.md]: Done

File: nix/sources.json (hunk 1 above)
Comment (nikolaglumac) [reply nix/sources.json:32]: Then we can not merge this today.

**Review by ManusMcCole — APPROVED** [review]
Goodjob @tomislavhoracek

**Comment by wadie914** [comment]
can somebody please give me a clear answer so basically none of the virgins
of the wallet can open my secret key file for now and they're working on
one that will is that my only solution?

**Pull request merged**